pyspark_ai/tool.py

```diff
@@ -104,3 +104,73 @@ def _run(
 
     async def _arun(self, *args: Any, **kwargs: Any) -> Any:
         raise NotImplementedError("ListTablesSqlDbTool does not support async")
+
+
+class VectorSearchUtil:
+    @staticmethod
+    def vector_similarity_search(
+        col_lst: list,
+        vector_store_path: Optional[str],
+        search_text: str,
+        col: str,
+        temp_name: str,
+    ):
+        from langchain.vectorstores import FAISS
+        from langchain.embeddings import HuggingFaceBgeEmbeddings
+        import os
+
+        if vector_store_path and os.path.exists(vector_store_path):
+            vector_db = FAISS.load_local(vector_store_path, HuggingFaceBgeEmbeddings())
+        else:
+            vector_db = FAISS.from_texts(col_lst, HuggingFaceBgeEmbeddings())
+
+            if vector_store_path:
+                vector_db.save_local(vector_store_path)
+
+        docs = vector_db.similarity_search(search_text)
+        return docs[0].page_content
+
+
+class SimilarValueTool(BaseTool):
+    """Tool for finding the semantically closest word to a keyword from a vector database."""
+
+    spark: Union[SparkSession, ConnectSparkSession] = Field(exclude=True)
+    name = "similar_value"
+    description = """
+    This tool finds the semantically closest word to a keyword from a vector database, using the FAISS library.
+    Input to this tool is a pipe-separated string in this format: keyword|column_name|temp_view_name.
+    The temp_view_name will be queried in the column_name for the semantically closest word to the keyword.
+    """
+
+    vector_store_dir: Optional[str]
+
+    def _run(
+        self, inputs: str, run_manager: Optional[CallbackManagerForToolRun] = None
+    ) -> str:
+        input_lst = inputs.split("|")
+
+        search_text = input_lst[0]
+        col = input_lst[1]
+        temp_name = input_lst[2]
+
+        new_df = self.spark.sql("select distinct `{}` from {}".format(col, temp_name))
+        col_lst = [str(row[col]) for row in new_df.collect()]
+
+        vector_store_path = (
+            (self.vector_store_dir + temp_name + "_" + col)
+            if self.vector_store_dir
+            else None
+        )
+
+        return VectorSearchUtil.vector_similarity_search(
+            col_lst, vector_store_path, search_text, col, temp_name
+        )
+
+    def _get_dataframe_results(self, df: DataFrame) -> list:
```

Comment:
This is duplicated with the code in pyspark_ai.py...let's move them to a new file spark_utils.py